File: javascript/go-counting/go-counting.js
//
// This is only a SKELETON file for the 'Go Counting' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class GoCounting {
  /**
   * Initialize the Go score calculator
   * @param {Array[string]} board - The go board to count a score from
   */
  constructor(board) {
    // Turn into a 2D array of string so we can mark claims when flood filling
    this.board = board.map((row) => row.split(''))
    this.height = board.length
    this.width = Math.min(...board.map((row) => row.length)) // if ragged array width is smallest width
    // . = claimed not processed
    // _ = claimed by none
    // b = claimed by black
    // w = claimed by white
    // W = white tile
    // B = black tile
    // ' ' = empty space on the board
    this.claimed = ['.', '_', 'b', 'w']
  }

  /**
   * Figure out who owns a region which includes a given point and what points
   * are included in that region
   * @param {Number} x x-coordinate of point to check
   * @param {Number} y y-coordinate of point to check
   * @returns {Object} Object with two fields, owner - who owns the territory, 
   * and territory an array of coordinates (presented as an array of length 2) 
   * belonging to the territory
   */
  getTerritory(x, y) {
    // Coordinate outside of the board
    if (x < 0 || x >= this.width || y < 0 || y >= this.height)
      return { error: 'Invalid coordinate' }

    // Black/White tiles cannot be flood-filled
    if (this.board[y][x] !== ' ')
      return { owner: 'NONE', territory: [] }

    let owner = this.floodFill(this.board, x, y, this.width, this.height)
    let results = this.processClaims(this.board, this.width, this.height)
    // clean up our mess - object may be reused.
    this.clearClaims(this.board, this.width, this.height)
    return {owner: owner, territory: owner === 'WHITE' ? results.territoryWhite : owner === 'BLACK' ? results.territoryBlack : results.territoryNone}
  }

  /**
   * Calculate all of the board locations owned by black, white and neither.
   * @returns {Object} Object with three fields. territoryBlack an array of coordinates
   * belonging to the black player (coordinates are a array of size 2 where x is at 0 
   * and y is at 1). territoryWhite the same but territory belonging to the white player,
   * and finally territoryNone coorinates owned by neither the black or white player.
   */
  getTerritories() {
    // flood fill everything
    for(let i = 0; i < this.width; i++)
    {
      for(let j = 0; j < this.height; j++)
      {
        if (this.board[j][i] === ' ')
          this.floodFill(this.board, i, j, this.width, this.height)
      }
    }
    let results = this.processClaims(this.board, this.width, this.height)
    // clean up our mess - object may be reused.
    this.clearClaims(this.board, this.width, this.height)
    return results
  }

  /**
   * Clear claimed sections of the board. Split out so it could be reused.
   * @param {Array[Array[string]]} board - 2D array with a string/character for each location on the board
   * @param {Number} width - Width of the board
   * @param {Number} height - Height of the board
   */
  clearClaims(board, width, height)
  {
    for(let i = 0; i < width;i++)
    {
      for(let j = 0; j < height; j++)
      {
        if (this.claimed.includes(board[j][i]))
        {
          board[j][i] = ' '
        }
      }
    }
  }

  /**
   * 
   * @param {Array[Array[string]]} board - 2D Array of strings where each string is a location on the board
   * @param {Number} x - x-coordinate of location to begin flood fill.
   * @param {Number} y - y-coordinate of location to begin flood fill.
   * @param {Number} width - width of the board.
   * @param {Number} height - height of the board
   * @returns {String} Who owns the given region, WHITE, BLACK, or NONE
   */
  floodFill(board, x, y, width, height) {    
    let black = false
    let white = false
    let stackCheck = []
    stackCheck.push([x, y])
    while (stackCheck.length > 0)
    {
      let [i, j] = stackCheck.pop()

      if (board[j][i] === ' ')
      {
        board[j][i] = '.'
        if ((j - 1 >= 0) && !this.claimed.includes(board[j - 1][i]))
          stackCheck.push([i, j - 1])
        if ((i - 1 >= 0) && !this.claimed.includes(board[j][i - 1]))
          stackCheck.push([i - 1, j])
        if ((j + 1 < height) && !this.claimed.includes(board[j + 1][i]))
          stackCheck.push([i, j + 1])
        if ((i + 1 < width) && !this.claimed.includes(board[j][i + 1]))
          stackCheck.push([i + 1, j])
      }
      else if (board[j][i] === 'B')
      {
        black = true
      }
      else if (board[j][i] === 'W')
      {
        white = true 
      }
    }

    let claim = '_'
    let owner = 'NONE'
    if (black && !white)
    {
      owner = 'BLACK'
      claim = 'b'
    }
    else if (white && !black)
    {
      owner = 'WHITE'
      claim = 'w'
    }

    for(let i = 0; i < width;i++)
    {
      for(let j = 0; j < height; j++)
      {
        if (board[j][i] === '.')
        {
          board[j][i] = claim
        }
      }
    }

    return owner
  }

  /**
   * Process all the claimed locations on the board.
   * @param {Array[Array[String]]} board - 2D Array of string where each string is a location on the board
   * @param {Number} width - Width of the board
   * @param {Number} height - Height of the board
   * @returns {Object} - Has three fields territoryBlack an array of coordinates (coordinates are a 2 entry array)
   * of all locations claimed by player Black. Likewise territoryWhite and territoryNone have an array of coordinates
   * claimed by player White, and player None. Coordinates should be in y then x sort order.
   */
  processClaims(board, width, height)
  {
    let black = [] 
    let white = []
    let unclaimed = []

    for(let i = 0; i < width;i++)
    {
      for(let j = 0; j < height; j++)
      {
        if (this.claimed.includes(board[j][i]))
        {
          let claimedBy = board[j][i]
          switch(claimedBy)
          {
            case '_':
              unclaimed.push([i, j])
              break;
            case 'w':
              white.push([i, j])
              break;
            case 'b':
              black.push([i, j])
          }
        }
      }
    }
    return {territoryBlack: black, territoryWhite: white, territoryNone: unclaimed}
  }
}

File: go/go-counting/go_counting.go
// Exercise: Counting Territories on a Go Board
package gocounting

import "errors"

// Holds points held by each player or neither
type AllTerritories struct {
	Black [][2]int
	White [][2]int
	None  [][2]int
}

// Game specific variables
type Game struct {
	board []string
}

// Create a new game object and return it
// @param board: Slice of strings where each string is a row of the board. The strings are assumed to be of length >= 1 and the same length on each row.
// @returns: Pointer to new Game struct.
func NewGame(board []string) *Game {
	return &Game{board: board}
}

// Territory - Compute the locations within the territory containing the point at x, y
// @param x: x-coordinate of point in territory
// @param y: y-coordinate of point in territory
// @returns: string is the owner of the territory ("BLACK", "WHITE", or "NONE"). The slice of int arrays of 2 elements are the points contained in the territory.
// @raises: Error is populated when the desired position is not on the board.
func (g *Game) Territory(x, y int) (string, [][2]int, error) {
	if x < 0 || y < 0 || y >= len(g.board) || x >= len(g.board[y]) {
		return "", make([][2]int, 0), errors.New("invalid coordinate")
	}
	temp := make([][]rune, len(g.board))
	for i := 0; i < len(g.board); i++ {
		temp[i] = []rune(g.board[i])
	}
	markers := map[rune]bool{'B': false, 'W': false, ' ': false, 'X': false}

	points := g.FloodFill(&temp, x, y, &markers)
	if markers['B'] && !markers['W'] {
		return "BLACK", points, nil
	} else if markers['W'] && !markers['B'] {
		return "WHITE", points, nil
	}
	return "NONE", points, nil
}

// Fill a region of the board and return the positions updated.
// @param board: Pointer to 2d rune array (mimics the board slice)
// @param x: x-coordinate of point to start at
// @param y: y-coordinate of point to start at
// @param markers: This holds the characters we bump into while populating the region. If you get no or both B and W it is none.
// @returns: slice of two element int arrays where each int array is a position populated by the flood fill.
func (g *Game) FloodFill(board *[][]rune, x, y int, markers *map[rune]bool) [][2]int {
	results := make([][2]int, 0)
	char := (*board)[y][x]
	if char == 'B' || char == 'W' || char == 'X' {
		return results
	}

	results = append(results, [2]int{x, y})
	(*board)[y][x] = 'X'
	// up
	up := [][2]int{}
	down := [][2]int{}
	left := [][2]int{}
	right := [][2]int{}

	if x-1 >= 0 {
		(*markers)[(*board)[y][x-1]] = true
		left = g.FloodFill(board, x-1, y, markers)
	}
	if x+1 < len((*board)[y]) {
		(*markers)[(*board)[y][x+1]] = true
		right = g.FloodFill(board, x+1, y, markers)
	}

	if y-1 >= 0 {
		(*markers)[(*board)[y-1][x]] = true

		up = g.FloodFill(board, x, y-1, markers)
	}
	if y+1 < len((*board)) {
		(*markers)[(*board)[y+1][x]] = true
		down = g.FloodFill(board, x, y+1, markers)
	}
	results = append(results, up...)
	results = append(results, down...)
	results = append(results, left...)
	results = append(results, right...)

	return results
}

// Similar to the Territory call but calculates all positions for all territories.
// @returns: AllTerritories struct where Black, White, and None are populated by points found by repeated
// flood fill calls on unpopulated areas over the whole board.
func (g *Game) Territories() AllTerritories {
	temp := make([][]rune, len(g.board))
	for i := 0; i < len(g.board); i++ {
		temp[i] = []rune(g.board[i])
	}

	var results AllTerritories = AllTerritories{}

	markers := map[rune]bool{'B': false, 'W': false, ' ': false, 'X': false}
	for y := 0; y < len(g.board); y++ {
		for x := 0; x < len(g.board[y]); x++ {
			if temp[y][x] == ' ' {
				markers['B'] = false
				markers['W'] = false
				points := g.FloodFill(&temp, x, y, &markers)
				if markers['B'] && !markers['W'] {
					results.Black = append(results.Black, points...)
				} else if markers['W'] && !markers['B'] {
					results.White = append(results.White, points...)
				} else {
					results.None = append(results.None, points...)
				}
			}
		}
	}
	return results
}
